cnd/src/network/protocol.rs

```diff
@@ -0,0 +1,163 @@
+        iter::once(b"/comit/swap/secret_hash/1.0.0")
+    }
+}
+
+impl<C> OutboundUpgrade<C> for OutboundProtocolConfig
+where
+    C: AsyncWrite + Unpin + Send + 'static,
+{
+    type Output = ();
+    type Error = Error;
+    type Future = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn upgrade_outbound(self, mut socket: C, info: Self::Info) -> Self::Future {
+        trace!(
+            "Upgrading outbound connection: {}",
+            String::from_utf8_lossy(info)
+        );
+        Box::pin(async move {
+            let bytes = serde_json::to_vec(&self.msg).unwrap();
+            upgrade::write_one(&mut socket, &bytes).await.unwrap();
+
+            Ok(())
+        })
+    }
+}
+
+/// Represents a prototype for an upgrade to handle the receiver side of the
+/// secret hash sharing protocol.
+#[derive(Clone, Copy, Debug)]
+pub struct InboundProtocolConfig;
+
+impl Default for InboundProtocolConfig {
+    fn default() -> Self {
+        Self {}
+    }
+}
+
+impl UpgradeInfo for InboundProtocolConfig {
+    type Info = &'static [u8];
+    type InfoIter = iter::Once<Self::Info>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        iter::once(b"/comit/swap/secret_hash/1.0.0")
+    }
+}
+
+impl<C> InboundUpgrade<C> for InboundProtocolConfig
+where
+    C: AsyncRead + Unpin + Send + 'static,
+{
+    type Output = Message;
+    type Error = Error;
+    type Future = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn upgrade_inbound(self, mut socket: C, info: Self::Info) -> Self::Future {
+        trace!(
+            "Upgrading inbound connection: {}",
+            String::from_utf8_lossy(info)
+        );
+        Box::pin(async move {
+            let message = upgrade::read_one(&mut socket, 1024).await?;
+            let mut de = serde_json::Deserializer::from_slice(&message);
+            let info = Message::deserialize(&mut de)?;
+
+            Ok(info)
+        })
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("read: ")]
+    Read(#[from] upgrade::ReadOneError),
+    #[error("deserialize: ")]
+    Deserialize(#[from] serde_json::Error),
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use futures::prelude::*;
+    use libp2p::core::{
+        multiaddr::multiaddr,
+        transport::{memory::MemoryTransport, ListenerEvent, Transport},
+        upgrade,
+    };
+    use rand::{thread_rng, Rng};
+    use std::str::FromStr;
+
+    #[tokio::test]
+    async fn correct_transfer() {
+        let mem_addr = multiaddr![Memory(thread_rng().gen::<u64>())];
+        let mut listener = MemoryTransport.listen_on(mem_addr).unwrap();
+
+        let send_msg = Message {
+            swap_id: SwapId::from_str("ad2652ca-ecf2-4cc6-b35c-b4351ac28a34").unwrap(),
+            secret_hash: SecretHash::from_str(
+                "bfbfbfbfbfbfbfbfbfbfbfbfbfbfbfbf\
+                 bfbfbfbfbfbfbfbfbfbfbfbfbfbfbfbf",
+            )
+            .unwrap(),
+        };
```

Comment:
The initialization order here is a bit weird because the message is in the middle of creating a listener and interacting with it.

```suggestion
		let send_msg = Message {
            swap_id: SwapId::from_str("ad2652ca-ecf2-4cc6-b35c-b4351ac28a34").unwrap(),
            secret_hash: SecretHash::from_str(
                "bfbfbfbfbfbfbfbfbfbfbfbfbfbfbfbf\
                 bfbfbfbfbfbfbfbfbfbfbfbfbfbfbfbf",
            )
            .unwrap(),
        };
        
        let mem_addr = multiaddr![Memory(thread_rng().gen::<u64>())];
        let mut listener = MemoryTransport.listen_on(mem_addr).unwrap();
```